diff --git a/utils/extract_symbols.py b/utils/extract_symbols.py
index 43f6039..5e09340 100755
--- a/utils/extract_symbols.py
+++ b/utils/extract_symbols.py
@@ -125,6 +125,13 @@ def readobj_is_32bit_windows(lib):
 # identifier/type mangling we can decide which symbols could possibly be
 # required and which we can discard.
 def should_keep_microsoft_symbol(symbol, calling_convention_decoration):
+    # required global variables
+    if (symbol == '?SetKey@?$AllAnalysesOn@VModule@llvm@@@llvm@@0UAnalysisSetKey@2@A'
+        or symbol == '?SetKey@?$AllAnalysesOn@VFunction@llvm@@@llvm@@0UAnalysisSetKey@2@A'
+        or symbol == '?AllAnalysesKey@PreservedAnalyses@llvm@@0UAnalysisSetKey@2@A'
+        or symbol == '?Key@?$InnerAnalysisManagerProxy@V?$AnalysisManager@VFunction@llvm@@$$V@llvm@@VModule@2@$$V@llvm@@0UAnalysisKey@2@A'):
+        return symbol
+
     # Keep unmangled (i.e. extern "C") names
     if not '?' in symbol:
         if calling_convention_decoration:
